# Sprint 3: Findings
---


In this notebook we will note out the findings and recommendations based on our analysis and modeling from Sprint 2 and Sprint 3.

In [5]:
#Loading necessary libraries
import pandas as pd

In [4]:

summary_model=pd.read_csv(r"C:\Users\karni\OneDrive\Desktop\GitHub\BrainStation_CapstoneProject\data\model_performance_summary.csv", index_col=0)
summary_model

,Train_MAE,Test_MAE,Train_MAPE,Test_MAPE
Mean,4.034347e+00,3.101882e+00,2.922166e-01,2.158319e-01
linreg_ohe,1.315639e+00,1.717609e+00,9.581909e-02,1.047659e-01
linreg_features,2.021206e-14,1.986290e-14,1.224942e-15,1.145575e-15
arima_111,3.826463e+00,5.577666e+00,2.330022e-01,3.979674e-01
arima_811,3.062481e+00,1.874828e+00,1.890975e-01,1.220753e-01
xgboost,4.008336e-04,1.089735e+00,2.342684e-05,5.827498e-02
xgboost_optimized,9.082587e-04,9.997100e-01,5.873678e-05,5.559898e-02
prophet,4.538755e+00,3.328523e+00,3.118428e-01,2.482262e-01


## Comparative Analysis of Model Performance

The table below summarizes the performance of eight predictive models on both training and test datasets using Mean Absolute Error (MAE) and Mean Absolute Percentage Error (MAPE). Lower values in these metrics indicate more accurate predictions.

| Model              | Train MAE | Test MAE | Train MAPE | Test MAPE |
|--------------------|-----------|----------|------------|-----------|
| Mean               | 4.03      | 3.10     | 0.29       | 0.22      |
| linreg_ohe         | 1.32      | 1.72     | 0.10       | 0.10      |
| linreg_features    | 0.00      | 0.00     | 0.00       | 0.00      |
| arima_111          | 3.83      | 5.58     | 0.23       | 0.40      |
| arima_811          | 3.06      | 1.87     | 0.19       | 0.12      |
| xgboost            | 0.00      | 1.09     | 0.00       | 0.06      |
| xgboost_optimized  | 0.00      | 1.00     | 0.00       | 0.06      |
| prophet            | 4.54      | 3.33     | 0.31       | 0.25      |




**Key Takeaways**

- The **linreg_features** model shows near-zero error on both train and test sets, indicating probable data leakage or target-feature overlap rather than true predictive power.
- **Optimized XGBoost** achieves the second lowest Test MAE (0.0010) and a competitive Test MAPE (5.56%), marking it as the most accurate overall.
- **Linear regression with one-hot encoding (linreg_ohe)** delivers decent generalization, with a Test MAE of 1.72 and Test MAPE of 10.48%.
- The **mean model** serves as a naive baseline; any model beating its 3.10 MAE and 21.58% MAPE offers genuine predictive value.
- **Traditional ARIMA and Prophet** time-series approaches lag behind tree-based methods in this context. 

## Model-by-Model Breakdown

### 1. Baseline “Mean” Model
- **Train MAE:** 4.03 &nbsp;&nbsp; **Test MAE:** 3.10  
- **Train MAPE:** 29.22% &nbsp;&nbsp; **Test MAPE:** 21.58%  
- Serves as a reference by predicting the historical average.

### 2. linreg_ohe (Linear Regression with One-Hot Encoding)
- **Train MAE:** 1.32 &nbsp;&nbsp; **Test MAE:** 1.72  
- **Train MAPE:** 9.58% &nbsp;&nbsp; **Test MAPE:** 10.48%  
- Solid performance; slight overfitting as test errors are higher than train.

### 3. linreg_features (Linear Regression on Engineered Features)
- **Train/Test Errors:** Near-zero (~10⁻¹⁴)  
- Improbably perfect fit suggests data leakage or target-feature overlap.

### 4. ARIMA Time-Series Models
- **arima_111:**  
    - **Test MAE:** 5.58 &nbsp;&nbsp; **Test MAPE:** 39.80%  
- **arima_811:**  
    - **Test MAE:** 1.87 &nbsp;&nbsp; **Test MAPE:** 12.21%  
- (8,1,1) configuration outperforms (1,1,1), benefiting from longer lag structure.

### 5. XGBoost Variants
- **xgboost:**  
    - **Test MAE:** 1.09 &nbsp;&nbsp; **Test MAPE:** 5.83%  
- **xgboost_optimized:**  
    - **Test MAE:** 1.00 &nbsp;&nbsp; **Test MAPE:** 5.56%  
- Hyperparameter tuning nearly eliminates absolute errors, showing strong pattern learning.

### 6. Prophet (Facebook’s Time-Series Library)
- **Train MAE:** 4.54 &nbsp;&nbsp; **Test MAE:** 3.33  
- **Train MAPE:** 31.18% &nbsp;&nbsp; **Test MAPE:** 24.82%  
- Robust for seasonality/trend, but underperforms compared to tuned tree-based models.

---





## Overfitting vs. Underfitting Insights

- **Overfitting Signals:**
    - linreg_ohe and ARIMA(8,1,1) show moderate train-test error gaps.
    - Perfectly zero errors in linreg_features strongly indicate data leakage.
- **Underfitting Signals:**
    - Mean model and Prophet have high errors, missing key data drivers.
    - ARIMA(1,1,1) underfits compared to more complex ARIMA/XGBoost variants.

---

# Next Steps:

- Based on the performance metrics, XGBoost optimized performs the best, so next steps would be to incorporate this into a predicted App (Currently I have a Streamlit Dashboard showing Energy Consumption trends)

- Drive into LSTM neural nets for forecasting energy consumption